src/services/identity/CollaboratorsService.ts

```diff
@@ -0,0 +1,240 @@
+        (
+          collaborator: User & {
+            SiteMember: SiteMember
+          }
+        ) => collaborator.id.toString() === requesterId,
+        // Prioritize the last logged in user
+        (
+          collaborator: User & {
+            SiteMember: SiteMember
+          }
+        ) => collaborator.lastLoggedIn,
+      ],
+      ["desc", "desc", "asc"]
+    )
+  }
+
+  create = async (siteName: string, email: string, acknowledged: boolean) => {
+    let site
+    let user
+
+    if (!email || !validator.isEmail(email)) {
+      return new BadRequestError(
+        "That doesn't look like a valid email. Try a gov.sg or other whitelisted email."
+      )
+    }
+
+    // 1. Check if email address is whitelisted, and derive the collaborator role
+    const derivedRole = await this.deriveAllowedRoleFromEmail(email)
+    if (!derivedRole) {
+      // Error - the user email is not whitelisted
+      logger.error(
+        `create collaborators error: user email ${email} is not whitelisted`
+      )
+      return new ForbiddenError(
+        `This collaborator couldn't be added. Visit our guide for more assistance.`
+      )
+    }
+
+    // 2. Check if site exists
+    site = await this.sitesService.getBySiteName(siteName)
+    if (!site) {
+      // Error - site does not exist
+      logger.error(`create collaborators error: site ${siteName} is not valid`)
+      return new NotFoundError(`Site does not exist`)
+    }
+
+    // 3. Check if valid user exists
+    user = await this.usersService.findByEmail(email)
+    if (!user) {
+      // Error - user with a valid gov email does not exist
+      logger.error(`create collaborators error: user ${email} is not valid`)
+      return new NotFoundError(
+        `This user does not have an Isomer account. Ask them to log in to Isomer and try adding them again.`
+      )
+    }
+
+    // 4. Check if user is already a site member
+    const existingSiteMember = await this.siteMemberRepository.findOne({
+      where: {
+        siteId: site.id,
+        userId: user.id,
+      },
+    })
+    if (existingSiteMember) {
+      return new ConflictError(`User is already a member of the site`)
+    }
+
+    // 5. Ensure that acknowledgement is true if the email role is contributor
+    if (derivedRole === CollaboratorRoles.Contributor && !acknowledged) {
+      return new UnprocessableError("Acknowledgement required")
+    }
+
+    // 6. Create the SiteMembers record
+    return this.siteMemberRepository.create({
+      siteId: site.id,
+      userId: user.id,
+      role: derivedRole,
+    })
+  }
+
+  delete = async (siteName: string, userId: string) => {
+    const site = await this.siteRepository.findOne({
+      where: { name: siteName },
+      include: [
+        {
+          model: User,
+          as: "site_members",
+        },
+      ],
+    })
+
+    const siteMembers = site?.site_members ?? []
+    const userIsSiteMember =
```

Comment:
Small nit: I think this can be changed to `isUserSiteMember` since it is a boolean?

Reply:
Agreed - amended in 92575fc